tests/python/relay/test_ir_module.py

```diff
@@ -0,0 +1,30 @@
+"""Unit tests for the Relay module."""
+import tvm
+from tvm import relay
+
+def test_global_mutual_recursion():
+    odd = relay.GlobalVar("odd")
+    even = relay.GlobalVar("even")
+
+    # even and odd are mutually recursive
+    x = relay.Var("x", relay.scalar_type('int32'))
```

Comment:
Strange thing @jroesch (I don't know who might be more knowledgeable about this): Type inference _fails_ if I don't annotate the argument type. I think this has to do with the relation for `equal` because the same thing happened when I tried to define a local function that used its argument in `equal`: it somehow can't conclude that `x` must be an int scalar even though the relation passes. Do you know what exactly could be going on there?

Reply:
`equal` uses `BroadcastCompRel` to generate type constraints, and it only generates type constraints if the type variables of two arguments are `TensorType`:
https://github.com/dmlc/tvm/blob/16b009b236de6dee49fac1d4d1dbc92d65b9ef8d/src/relay/op/type_relations.cc#L113-L114

I tried to provide a `relay.TypeVar` instead of concrete tensor type as the line did, it won't add type constraint. Probably we can relax `ConcreteBroadcast` to deal with `TypeVar` or `IncompleteTypeNode`?

Reply:
With the following change, type inference works without annotation. 
````bash
(3.6.3.py) ➜  tvm git:(relay-mutual-recursion) ✗ git diff src/relay/op/type_relations.cc
--- i/src/relay/op/type_relations.cc
+++ w/src/relay/op/type_relations.cc
@@ -110,6 +110,7 @@ bool BroadcastCompRel(const Array<Type>& types,
   CHECK_EQ(types.size(), 3);
   RELAY_LOG(INFO) << "In1: " << types[0] << "In2: " << types[1]
                   << "Out: " << types[2] << std::endl;
+  reporter->Assign(types[0], types[1]);
   if (auto t0 = ToTensorType(types[0])) {
     if (auto t1 = ToTensorType(types[1])) {
````

Reply:
Thanks for taking a look. @jroesch seemed to imply in recent discussions that a larger rewrite of these type relations to correctly handle incomplete types and type vars will have to be done in the not-very-distant future

Reply:
I should additionally state that I wonder whether changes to relations should be in a separate PR or part of a larger push, since the changes Jared envisions probably apply to many places.

Reply:
Will there be any RFCs? @jroesch  @slyubomirsky